File: python/242. Valid Anagram.py
class Solution:
    def isAnagram(self, s: str, t: str) -> bool:
        # This solution is O(nlogn)
        # return sorted(s) == sorted(t)

        # This solution is O(n) and involves using a bucket array to keep track of the characters

        bucket = [0] * 26

        for c in s:
            bucket[ord(c) - ord('a')]+=1
        for c in t:
            bucket[ord(c) - ord('a')]-=1

        for n in bucket:
            if n != 0: return False
        
        return True
            
        # Another similar solution implementing a map
        # map = {}
        
        # for c in s:
        #     if c not in map:
        #         map[c] = 0
        #     map[c]+=1
        # for c in t:
        #     if c not in map:
        #         map[c] = 0
        #     map[c]-=1

        # for n in map.values():
        #     if n != 0: return False
    
        # return True

File: java/242. Valid Anagram.java
//Neetcode -> Array & Hashing

class Solution {
    public boolean isAnagram(String s, String t) {
       //if they're not the same length then they can't be anagrams 
        if (s.length() != t.length()){
            return false;
        }
        
        //bucket array to keep track of ASCII lower case alphabets/ characters 
        int[] bucket = new int[26];
        
        //incrementing index value of each character as we find them in our first string
        for (int i=0; i<s.length();i++){
            bucket[s.charAt(i) - 'a']++;
        }
        
        //decrementing index value of each character as we find them in our second string
        for (int i=0; i<t.length();i++){
            bucket[t.charAt(i) - 'a']--;
        }
        
        //if we have zero for every index in our bucket[] then it means every character in "s" was present in "t"
        for (int n: bucket){
            if (n != 0){
                return false;
            }
        }
        return true;
    }
}
